### PR TITLE
Fixed #26834 -- Included min/max values in IntegerField formfield

### PR DESCRIPTION
**Pull request by rixx**
Rebased from #6964.

https://code.djangoproject.com/ticket/26834

Edit: Tests are breaking, I'll look into it.

### CODE REVIEW

**Comment by felixxm** [comment]
@rixx Thanks! Some test are falling there is also an issue with multiple validators (see [comment](https://github.com/django/django/pull/6964#issuecomment-234648745)), and mysterious "issue related to the validators property" (see [comment](https://github.com/django/django/pull/6964#issuecomment-237743504)).

**Comment by rixx** [comment]
Ah, I had unfairly ignored the issues with multiple validators. I added handling and tests for this case.

As for the validators property, I believe it's suitable for this feature, so I don't see any issue with it at this point.

**Comment by rixx** [comment]
Tests are breaking, and this is why: On databases with proper limits on their integer fields (so all supported ones apart from sqlite3), there are Min/Max validators on IntegerFields to prevent overflows and keep IntegerFields to the values they are rated for.

This would now result in these min/max values to be added to form fields, e.g. -2147483648 as a typical min value on a form with no further min value validation.

I'd argue that this is actually well-behaved form generation, as the form *should* prevent values from reaching the database that we know will now be acceptable. But it makes this change more frontend facing. Is this acceptable? If it is, I'll need some minor help in building better tests and how to phrase the release notes.

**Comment by pope1ni** [comment]
Also note that `limit_value` can be callable as of 2.2 which required a fix released in 2.2.1 - see https://github.com/django/django/pull/11180.

I would recommend revisiting my suggestion in that PR to make `limit_value` a settable property to hide away the callable stuff so that we can avoid the potential for crashes in all new places that attempt to use `limit_value`.

Note that my suggestion was rejected at the time as too invasive for a regression fix, but this highlights the exact situation my proposal was aiming to prevent.

**Comment by rixx** [comment]
Hmm, so either we wait for your suggestion to be implemented, or we add another piece of code and tests for callable values here. For the little it's worth, I fully agree with your suggested property based API – having all places that interact with Min/Max validators check for `callable(my_validator.limit_value)` seems unwieldy.

**Comment by pope1ni** [comment]
> ...having all places that interact with Min/Max validators check for `callable(my_validator.limit_value)` seems unwieldy.

Especially when the result of forgetting to do so is a crash.

I have prepared some code to refactor `limit_value` to handle callables transparently and will try to get a PR together later today if possible.

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -1833,6 +1833,27 @@ def to_python(self, value):
...
+            if isinstance(validator, validators.MinValueValidator):
+                value = (
+                    validator.limit_value()
+                    if callable(validator.limit_value)
+                    else validator.limit_value
+                )
+                if 'min_value' in kwargs:
+                    kwargs['min_value'] = max(value, kwargs['min_value'])
+                else:
+                    kwargs['min_value'] = value
```
Comment (pope1ni) [inline django/db/models/fields/__init__.py]:
```python
kwargs['min_value'] = max(value, kwargs.get('min_value', value))
```

File: django/db/models/fields/__init__.py [hunk 2]
```diff
@@ -1833,6 +1833,27 @@ def to_python(self, value):
...
+            elif isinstance(validator, validators.MaxValueValidator):
+                value = (
+                    validator.limit_value()
+                    if callable(validator.limit_value)
+                    else validator.limit_value
+                )
+                if 'max_value' in kwargs:
+                    kwargs['max_value'] = min(value, kwargs['max_value'])
+                else:
+                    kwargs['max_value'] = value
```
Comment (pope1ni) [inline django/db/models/fields/__init__.py]:
```python
kwargs['max_value'] = min(value, kwargs.get('max_value', value))
```

**Comment by rixx** [comment]
Yes, I think we should figure out first if

- we are comfortable changing pretty much all forms with `IntegerField`s, potentially breaking many tests out there, and
- potentially leaking the database in use by providing database-related min/max values

before I continue working on this.

**Comment by felixxm** [comment]
@rixx Sorry for a late response :disappointed: . I missed your comment.

We should reach a wider audience and consensus on [the mailing list](https://groups.google.com/g/django-developers) before moving this forward.

Closing for now.

**Pull request closed without merging**